File: kmeans.py
import math
import sys

def kMeans(dataPoints, k, n, d, iter=200 ):
    # initialize centroids to first k data points
    epsilon = 0.001
    clusters = []
    centroids = []
    for i in range(k):
        centroids.append({'center': dataPoints[i].copy(), 'size': 1, 'lastDelta': 1 + epsilon})
        clusters.append(i)
    
    for i in range(k,n):
        clusters.append(-1)

    i = 0
    maxDelta = epsilon + 1
    while(i < iter and maxDelta > epsilon ):
        oldCluster = clusters[i%n]
        closestCluster = findClosestCluster(dataPoints[i%n], centroids)
        
        clusters[i%n] = closestCluster
        
        if(oldCluster == -1):
            centroids[closestCluster] = addDataPoint(centroids[closestCluster], dataPoints[i%n])  
            
        if(oldCluster != -1 and oldCluster != closestCluster):
            centroids[closestCluster] = addDataPoint(centroids[closestCluster], dataPoints[i%n])  
            centroids[oldCluster] = removeDataPoint(centroids[oldCluster], dataPoints[i%n])
            
        maxDelta = 0
        for u in centroids:
            if(u['lastDelta'] > maxDelta):
                maxDelta = u['lastDelta']
        i += 1
    print(centroids)
    print(i)

def findClosestCluster(datapoint, centroids):
    minDistance = calcEclideanDistance(centroids[0]['center'], datapoint)
    closestCluster = 0
    for index, u in enumerate(centroids):
        distance = calcEclideanDistance(u['center'], datapoint)
        if(distance < minDistance):
            minDistance = distance
            closestCluster = index
    return closestCluster

def calcEclideanDistance(u, v):
    squareSum = 0
    for i in range(len(u)):
        squareSum += (u[i] - v[i])*(u[i] - v[i])
    return math.sqrt(squareSum)
    

def addDataPoint(centroid, datapoint):
    oldCenter = centroid['center'].copy()
    for index, w in enumerate(centroid['center']):
        # ( x - 1 / x ) * oldAverage + ( 1 / x ) * newElement
        centroid['center'][index] = (centroid['size'] * w + datapoint[index])/(centroid['size'] + 1)
    newDelta = calcEclideanDistance(oldCenter, centroid['center'])
    centroid['size'] +=1
    centroid['lastDelta'] = newDelta
    return centroid

def removeDataPoint(centroid, datapoint):
    oldCenter = centroid['center'].copy()
    for index, w in enumerate(centroid['center']):
        centroid['center'][index] = ((w*centroid['size']) - datapoint[index])/(centroid['size']-1)
    newDelta = calcEclideanDistance(oldCenter, centroid['center'])
    centroid['size'] -=1
    centroid['lastDelta'] = newDelta
    return centroid

if __name__ == '__main__':
    if(len(sys.argv) < 5):
        print("An Error Has Occurred")
        exit
    try:
        k = int(sys.argv[1])
        n = int(sys.argv[2])
        d = int(sys.argv[3])
        if len(sys.argv == 5):
            iter = 200
            filePath = sys.argv[4]
        if len(sys.argv) == 6:
            iter = int(sys.argv[4])
            filePath = sys.argv[5]
        
            
        kMeans(list, k, n, d)
    except:
        print("An Error Has Occurred")

File: kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define EPSILON 0.001

typedef struct
{
   int size;
   float *center;
   float *currentCenter;
} Centroid;

int findClosestCluster(float[], Centroid[], int, int);
float CalcEclideanDistance(float[], float[], int);
int isNumber(char[]);

float CalcEclideanDistance(float u[], float v[], int d)
{
   float squareSum;
   int i;
   squareSum = 0;
   for (i = 0; i < d; i++)
   {
      squareSum += (u[i] - v[i]) * (u[i] - v[i]);
   }

   return sqrt(squareSum);
}

int findClosestCluster(float dataPoint[], Centroid centroids[], int d, int k)
{
   int closest, i;
   float minDistance, distance;
   closest = 0;
   minDistance = CalcEclideanDistance(dataPoint, centroids[0].center, d);

   for (i = 0; i < k; i++)
   {
      distance = CalcEclideanDistance(dataPoint, centroids[i].center, d);
      if (distance < minDistance)
      {
         minDistance = distance;
         closest = i;
      }
   }
   return closest;
}

int isNumber(char number[])
{
   int i;
   i = 0;

   if (number[0] == '-')
      i = 1;
   for (; number[i] != 0; i++)
   {

      if (number[i] > '9' || number[i] < '0')
         return 0;
   }
   return 1;
}

int main(int argc, char *argv[])
{
   /* declerations */
   int i, j, l, n, k, d, iter, closest;
   float current, maxDelta, delta;
   Centroid *centroids;
   float **dataPoints;

   /* parsing parameters */
   if (!(argc == 4 || argc == 5))
   {
      printf("An Error Has Occurred");
      exit(EXIT_FAILURE);
   }

   k = atoi(argv[1]);
   n = atoi(argv[2]);
   d = atoi(argv[3]);
   iter = 200;
   if (argc == 5)
   {
      iter = atoi(argv[4]);
      if ((iter <= 0 || iter >= 1000) || !isNumber(argv[4]))
      {
         fprintf(stderr, "Invalid maximum iteration!");
         exit(EXIT_FAILURE);
      }
   }
   if (n <= 1 || !isNumber(argv[2]))
   {
      fprintf(stderr, "Invalid number of points!");
      exit(EXIT_FAILURE);
   }
   if ((k <= 1 || k >= n) || !isNumber(argv[1]))
   {
      fprintf(stderr, "Invalid number of clusters!");
      exit(EXIT_FAILURE);
   }
   if (d <= 0 || !isNumber(argv[3]))
   {
      fprintf(stderr, "Invalid dimension of point!");
      exit(EXIT_FAILURE);
   }

   /* allocation */

   centroids = calloc(k, sizeof(Centroid));
   dataPoints = (float **)calloc(n, sizeof(float *));

   if (centroids == NULL || dataPoints == NULL)
   {
      fprintf(stderr, "An Error Has Occurred");
      exit(EXIT_FAILURE);
   }

   for (i = 0; i < n; i++)
   {
      dataPoints[i] = (float *)calloc(d, sizeof(float));
      if (dataPoints[i] == NULL)
      {
         fprintf(stderr, "An Error Has Occurred");
         exit(EXIT_FAILURE);
      }
   }

   /* parsing the input */
   for (i = 0; i < n; i++)
   {
      for (j = 0; (j < d) && (scanf("%f,", &current) != EOF); j++)
      {
         dataPoints[i][j] = current;
      }
   }

   /* initialization and centroids allocation */
   for (i = 0; i < k; i++)
   {
      centroids[i].size = 0;
      centroids[i].center = (float *)calloc(d, sizeof(float));
      centroids[i].currentCenter = (float *)calloc(d, sizeof(float));
      if (centroids[i].center == NULL || centroids[i].currentCenter == NULL)
      {
         fprintf(stderr, "An Error Has Occurred");
         exit(EXIT_FAILURE);
      }

      for (j = 0; j < d; j++)
      {
         centroids[i].center[j] = dataPoints[i][j];
         centroids[i].currentCenter[j] = 0;
      }
   }

   /* main loop */
   maxDelta = 1 + EPSILON;
   for (i = 0; (i < iter) && (maxDelta > EPSILON); i++)
   {
      /* find closest cluster for every x in dataPoints */
      for (j = 0; j < n; j++)
      {
         closest = findClosestCluster(dataPoints[j], centroids, d, k);
         centroids[closest].size += 1;
         for (l = 0; l < d; l++)
         {
            centroids[closest].currentCenter[l] = centroids[closest].currentCenter[l] + dataPoints[j][l];
         }
      }

      /* recenter all clusters and update maxDelta */
      maxDelta = 0;
      for (j = 0; j < k; j++)
      {
         for (l = 0; l < d; l++)
         {
            centroids[j].currentCenter[l] = (centroids[j].currentCenter[l]) / (centroids[j].size);
         }
         delta = CalcEclideanDistance(centroids[j].currentCenter, centroids[j].center, d);
         if (delta > maxDelta)
         {
            maxDelta = delta;
         }

         for (l = 0; l < d; l++)
         {
            centroids[j].center[l] = centroids[j].currentCenter[l];
            centroids[j].currentCenter[l] = 0;
            centroids[j].size = 0;
         }
      }
   }

   /* print out the centrouds */
   for (i = 0; i < k; i++)
   {
      for (j = 0; j < d; j++)
      {
         printf("%.4f,", centroids[i].center[j]);
      }
      printf("\n");
   }

   /* free all allocated space */
   for (i = 0; i < k; i++)
   {
      free(centroids[i].center);
      free(centroids[i].currentCenter);
   }
   free(centroids);

   for (i = 0; i < n; i++)
   {
      free(dataPoints[i]);
   }
   free(dataPoints);
   return 0;
}